### DISCUSSION TITLE
Implementing the Framer/Deframer with a Serial Driver

### DISCUSSION THREAD

**Discussion by ebenezerarunkumar** [nasa/fprime#1312] [Q&A] [answered]
Currently the Framer and Deframer use the port types `Drv::ByteStreamSend` `Drv::ByteStreamRecv` and `Drv::ByteStreamPoll` to send and receive data from `ByteStreamDriverModel` components like the TcpClient, TcpServer, and Udp. 

I am currently working on a deployment for an ATmega128 that needs to use a Serial Driver instead of the TcpClient. However, I am unsure how to correctly send data between the serial driver and the Framer/Deframer, since the serial driver uses the `Drv::SerialWrite` and `Drv::SerialRead` port types. Is there any way to configure my topology without having to edit the framer and deframer components? If not, would I just have to change the framer and deframer to use the `Drv::SerialWrite` and `Drv::SerialRead` port types?

An alternative strategy would be to rework the serial driver component to be `ByteStreamDriverModel` compliant...

**Comment by LeStarch** [comment, marked as answer]
There are two options I would suggest:

1. Rework the Serial driver to be `ByteStreamDriverModel` compliant
2. Create an adapter component from `ByteStreamDriverModel` => `Drv::Serial`

I would not touch the framer/deframer.

**Reply by LeStarch** [reply]
Item 2 is fairly straightforward.  It would be a `passive` component that implements the `ByteStreamDriverModel` interface and delegates to `SerialRead` and `SerialWrite` where appropriate.

**Reply by LeStarch** [reply]
You don't have to implement both Poll and Revc ports of `ByteStreamDriverModel`.  Your SerialDriver will expect to be polled or have a thread receive and supply data through a read port. 

Once done, you can interpose this adapter between framer/deframer and the serial driver.

**Reply by ebenezerarunkumar** [reply]
thanks! this seems simple enough!
